service/paxos/paxos_state.hh

```diff
@@ -49,9 +49,7 @@ private:
         bool _locked = false;
     public:
         future<> lock() {
```

Comment:
Could the `guard` be destroyed while `lock` is in progress?

I don't like this design in general, it's easy to make mistakes. `guard` should represent **already locked** state, i.e. there should be a function that takes the semaphore and returns a `guard`, here we're doing things in opposite order than it should be, first create a `guard` which is actually not yet functioning as a guard at this stage, then we `lock` which makes it a real guard. And now we have to worry about having it moved or destroyed before `lock` finishes, and in general worry about 3 states: not locked, locking in progress, already locked.

Here's how I think `guard` should look like:
```cpp
class guard {
    key_lock_map& _map;
    dht::token _key;

    guard(key_lock_map& map, const dht::token& key) : _map(map), _key(key) {};

public:
    guard(guard&& o) noexcept = default;

    ~guard() {
        _map.get_semaphore_for_key(_key).signal(1);
        _map.release_semaphore_for_key(_key);
    }

    friend future<guard> make_guard(key_lock_map& map, dht::token key, clock_type::time_point timeout) {
        co_await map.get_semaphore_for_key(key).wait(timeout, 1);
        co_return guard{map, key};
    }
}
```

Note that the constructor is private, the only way to construct it (other than move) is `make_guard` free-function.

Safe by design -- impossible to move/destroy it before lock is taken because it doesn't exist yet. The original bug also becomes impossible to write. Fewer variables to store too, less state (no `_locked`). Code is shorter.
It's just strictly better on all angles.

Implementations of `get_replica_lock` and `get_cas_lock` also become shorter. They just `return guard::make_guard(...)`.

Reply:
Ok, looking at the use cases, the guard never needs to be moved I think -- does `co_return guard{...}` need to move it? If not, then we can delete the move constructor and then it's safe